### DISCUSSION TITLE
Trying to apply custom .onnx model

### DISCUSSION THREAD

**Discussion by Jeatable** [AICAN-Research/FAST-Pathology#100] [Q&A]
Hi all.

I'm trying to apply an .onnx model with the Windows build. Initially placed it in the /fastpathology/models folder with an identical named .txt, but didn't find a possiblity to apply it in FASTPathology.

So I downloaded one of the existing colon HE models, and mimicked its structure. Created a title-model and title-fp folder in /fastpathology/datahub, placed my .onnx file in the -model folder and changed PipelineName, Attribute classes and Attribute model in the colon HE pipeline.fpl file and saved it to the -fp folder.

Two warnings in fastpahtology but it seems to be running,
WARNING [16048] Requested magnification level does not exist in image pyramid. Will now try to sample from a lower level and resize. This may increase runtime.
WARNING [16048] Patch size must be a multiple of 16 (TIFF limitation). Adding some overlap (32, 32) to fix.

However it has been running for 5 hours now, with the progress bar still at 0%, but definitly with activity on the GPU. Is this normal during auto-tuning, or should I edit the .fpl file to prevent the warnings and restart?

**Comment by andreped** [comment]
> WARNING [16048] Requested magnification level does not exist in image pyramid. Will now try to sample from a lower level and resize. This may increase runtime.

There are several things that might be causing the slow runtime, but I think this is one of them. When you create a pyramid from a non-pyramid image, there is no magnification level stored. By default, if no magnification level is provided, I believe FAST assumed it to be 40 (as in x40). In your case, what did you set as magnification? Which resolution would you expect to run your model on? I assume it was trained on a lower magnification plane like 10x?

Could you share the corresponding FPL file, and I could have a look?

I assume you are following this tutorial (see https://github.com/andreped/NoCodeSeg) or potentially this video (see https://www.youtube.com/watch?v=9dTfUwnL6zY)?

**Reply by Jeatable** [reply]
Yes, that's the tutorial I followed! :]

The dataset I use is imaged at 4x, however in the metadata of the original images magnification is stated as 0.50. I applied downsample 2 for export from QuPath to MIB, so following the tutorial I figured magnification should be 2. My goal is in the end to analyse shape irregularities in segmented objects so I dont want to downsample too much.

I renamed the .fpl to .txt for attaching it here.
[pipeline.txt](https://github.com/user-attachments/files/16309343/pipeline.txt)

**Reply by andreped** [reply]
I see, 2x is not really a standard magnification level. But what happens if you just set `40` as the magnification level in the pipeline? In that case, the full resolution of 2x may be used - don't recall this logic entirely. Alternatively, just avoid setting it, and hopefully the full resolution should be used. Could you give it a try?

To make debugging easier I pasted the FPL below - it also makes it easier for others finding this thread:
```
PipelineName "Agg segmentation monochrome no stain"
PipelineDescription "Segmentation of Agg in aggarose from bioreactors, based on https://www.frontiersin.org/articles/10.3389/fmed.2021.816281/full - https://github.com/andreped/NoCodeSeg."

PipelineInputData WSI "Whole-slide image"
PipelineOutputData segmentation stitcher 0
Attribute classes "Background;Positive"

### Processing chain
ProcessObject tissueSeg TissueSegmentation
Attribute threshold 70
Input 0 WSI

ProcessObject patch PatchGenerator
Attribute patch-size 512 512
Attribute patch-magnification 2
Attribute patch-overlap 0.05
Attribute mask-threshold 0.02
Input 0 WSI
Input 1 tissueSeg 0

ProcessObject network SegmentationNetwork
Attribute scale-factor 1.0
Attribute model "/path/to/some/model.onnx"
Input 0 patch 0

ProcessObject stitcher PatchStitcher
Input 0 network 0

### Renderers
Renderer imgRenderer ImagePyramidRenderer
Input 0 WSI

Renderer segRenderer SegmentationRenderer
Attribute opacity 0.5
Attribute border-opacity 1.0
Input 0 stitcher 0
```

**Reply by Jeatable** [reply]
I tried `40` which gives the same two warnings. Some GPU activity but after half an hour still at 0% progress in FASTpathology. Now I'm trying with excluding the `Attribute patch-magnification` line altoghether. No warnings at all, just the OK parsing, but no percentage displayed at all in FASTpathology, and funny GPU activity, core and memory clocks ramping up, but virtualy no load on either core or memory, while in all previous cases with warnings they were around 20-30%.

**Reply by andreped** [reply]
@Jeatable Would it be possible to share a model and an image you are using for testing, to reproduce the issue

 You can send it on my email, if you don't want to make these available to anyone else: andrped94@gmail.com

**Reply by Jeatable** [reply]
I emailed you an image, and the model.

On a side note, I checked if the model still worked on the vips generated pyramid files in QuPath/deepMIB. Which wasn't the case for jpeg compressed. On no compression, or deflated (which I sent you) it generated the result I expected in less then a minute.

**Reply by andreped** [reply]
I can have a look later today, @Jeatable :]

**Reply by Jeatable** [reply]
I'm curious what you'll find!

Getting a solution to only use FAST pathology and the .onnx model to obtain bitmaps would be preferred, esspecially for the less script savy users in my group. However for the time being I found a workaround, modifying the latest exportingtiles groovy script used in the NoCodeSeg pipeline: 

line 30 Removed `int patchSize` line altogether
line 31 Removing this line broke script so set `int pixelOverlap` to `= 0` 
line 47 Changed `.backgroundLabel(0, ColorTools.WHITE)` to `.backgroundLabel(255, ColorTools.BLACK)` 
line 65 Changed `.tileSize(patchSize)` to full image dimensions, for me `.tileSize(10240, 9216)` 

Last assumably also possible with setting variables but I didn't tried for too long, since this works.
